Add atime and mtime support for files ops, start with files.put

vram0gh2 opened the pull request:
Add concrete support for ``atime`` and ``mtime`` checking/setting in the ``files.put`` operation and provide infrastructure to add this time metadata support to other operations as well.

Includes a fix in ``facts/files.py`` to accommodate negative timestamps from ``stat`` on the remote host.

Added support for more ``os.stat_result`` file metadata to be provided in the test json files, by updating the mock ``os.stat`` patch used for testing.  This includes ``atime``, ``mtime``, but also the other members of the standard ``stat`` 10-tuple as well.

@Fizzadar: Looking for any feedback/implementation concerns, style changes you'd like, and how best to handle documenting the specifics and caveats.  I added a note in the operation docstring, but if ``atime`` and ``mtime`` support is implemented in other operations (*e.g.*, ``files.file``) these will be true there as well.  Is there a preferred way to break that information out into something outside of the function-specific docstring?

The expanded mock ``os.stat`` added a fair number of LOC, but it's also now equipped for future features/higher fidelity.

Also, I used the American spelling of "normalize" in one of the added support functions, happy to change it to British if preferred.

- [x] Pull request is based on the default branch (`3.x` at this time)
- [x] Pull request includes tests for any new/updated operations/facts
- [x] Pull request includes documentation for any new/updated operations/facts
- [x] Tests pass (see `scripts/dev-test.sh`)
- [x] Type checking & code style passes (see `scripts/dev-lint.sh`)

vram0gh2 committed:
Add file_utils.touch(), prep for file ops atime and mtime metadata support

Add a restricted analog to the ``touch`` command.  Supports mutating atime and
mtime, however only allows modification of one time field at a time.  A single
unambiguous datetime val is the time source since any ``touch`` operation with
specified atime/mtime requires the month, day, and time at a minimum.  Is BSD
aware in that datetimes are always converted to UTC rather than relying on the
richer timezone support in GNU ``touch``.

The single time field setting approach is less flexible per invocation than
native ``touch``, but also far simpler and nicely fits with file operations and
their way of working on a single attribute at a time.

vram0gh2 committed:
Add support for atime and mtime in files.put()

Time metadate updates are not coalesced, as ``touch`` can do with a single date
and time arg, or via reference files.  However, this actually allows more
flexibility than native ''touch`` by allowing both atime and mtime values to be
specified independently.

Two support functions are added as well.

vram0gh2 committed:
Rewrite mock os.stat in tests/util.py and apply to pyinfra.operations.files

In preparation for tests of mtime and ctime support, make the mock os.stat
actually use any stat fields provided in the test json, with non-zero defaults
for all fields.

Doesn't break any existing test, but lays foundation for any future ops or facts
which may use this info.

Doesn't currently support setting the st_{a, m}time_ns high-resolution time
fields.

vram0gh2 committed:
Fix the stat regex in facts/files.py to allow negative timestamps

Negative timestamps are valid and permissible.  Ensure
pyinfra.facts.files.stat doesn't fail to match stat command output when such
values occur.

vram0gh2 committed:
Change test/utils.py parse_value() to use datetime.fromisoformat

Change the datetime Python value to be generated from the fromisoformat method
rather than strptime.  This yields identical results as before when no
UTC offset information is provided, but respects it if it does.  It is also
consistent with the os.stat parsing.

vram0gh2 committed:
Add tests for ops/files.put atime and mtime support

Add several testcases to cover multiple aspects of functionality:
- Change detection, incl. differences <1s that are visible in the seconds vals
- Separate actions on mtime and atime
- Local files as the time references
- Remote files as the time references
- Use of the server's timezone when a datetime arg has no tzinfo

Fizzadar commented on pyinfra/operations/files.py:
```diff
@@ -778,6 +778,56 @@ def get(
 
 
+def _canonicalize_timespec(field, local_file, timespec):
+    assert field == "atime" or field == "mtime"
```
Using an enum for the field argument would be cleaner here + support type-checking.

Fizzadar commented on pyinfra/operations/files.py:
```diff
@@ -778,6 +778,56 @@ def get(
+            return timespec
+    elif isinstance(timespec, bool) and timespec:
+        lf_stat = os.stat(local_file)
+        if field == "atime":
+            return datetime.fromtimestamp(lf_stat.st_atime, tz=timezone.utc)
+        else:
+            return datetime.fromtimestamp(lf_stat.st_mtime, tz=timezone.utc)
```
```suggestion
        tstamp = if_stat.st_atime if field == "atime" else if_stat.st_mtime
        return datetime.fromtimestamp(tstamp, tz=timezone.utc)
```

Nitpicking, just a bit neater.

Fizzadar commented on pyinfra/operations/files.py:
```diff
@@ -802,6 +854,9 @@ def put(
     + assume_exists: whether to assume the local file exists
+    + atime: value of atime the file should have, use ``True`` to match the local file
+    + mtime: value of atime the file should have, use ``True`` to match the local file
+    + timesrc: the source of the time value if atime or mtime are ``True``
```
Leftover argument?

Fizzadar reviewed (commented):
Nice, this is really neat @vram0gh2! Sorry it took so long for me to do a proper review. Replying to your notes specifically:

> I added a note in the operation docstring, but if atime and mtime support is implemented in other operations (e.g., files.file) these will be true there as well. Is there a preferred way to break that information out into something outside of the function-specific docstring?

Currently not, and yeah that could be messy to duplicate everywhere. One option would be to put it into the top level docstring in `operations/files.py` such that it appears at the top of the [operation doc](https://docs.pyinfra.com/en/3.x/operations/files.html) and then reference that within each operation.

I do think this would be a great to expand to the other ops!

> The expanded mock os.stat added a fair number of LOC, but it's also now equipped for future features/higher fidelity.

👍 really nice, good add.

> Also, I used the American spelling of "normalize" in one of the added support functions, happy to change it to British if preferred.

All good, American makes sense as consistent w/Python, though I'm sure pyinfra contains a mix of both already 🙃

vram0gh2 committed:
Remove extraneous dangling argument docstring leftover from prior attempt.

vram0gh2 committed:
Use enum as flag for atime and mtime in internal functions.

In operations/util/files.py, define an enum to indicate whether the metadata
time field being manipulated by touch() is atime or mtime.`

This feeds into operations/files.py and the _canonicalize_timespec() support
function.

vram0gh2 committed:
Fix a copypasta miss in the docstring for the mtime arg.

vram0gh2 committed:
Simplify a conditional block in operations/files.py:_canonicalize_timespec()

vram0gh2 commented:
Thanks for the review, @Fizzadar !

I made changes/fixes as per the feedback .  Docstrings should now be fixed, and we're using an ``enum`` to indicate ``atime`` or ``mtime``.

There could be some stylistic room to argue in that I gave the enum a somewhat? long name of ``MetadataTimeField`` to be clear it wasn't referring to a portion of an ISO date string or components of a ``datetime``.  This results in a few spots where lines are wrapped.  I wouldn't object to shortening this if the ambiguity argument is outweighed by the desire for compactness.

If this is considered merge-ready otherwise, perhaps we can address the documentation issue if/when ``atime`` and ``mtime`` support is added to other ops and leave it where it is for now.

Fizzadar commented:
> Thanks for the review, @Fizzadar !
> 
> I made changes/fixes as per the feedback . Docstrings should now be fixed, and we're using an `enum` to indicate `atime` or `mtime`.

🙏 thank you!

> There could be some stylistic room to argue in that I gave the enum a somewhat? long name of `MetadataTimeField` to be clear it wasn't referring to a portion of an ISO date string or components of a `datetime`. This results in a few spots where lines are wrapped. I wouldn't object to shortening this if the ambiguity argument is outweighed by the desire for compactness.

I think the name is good as-is, rather a more explicit name than something shorter for styles sake!

> If this is considered merge-ready otherwise, perhaps we can address the documentation issue if/when `atime` and `mtime` support is added to other ops and leave it where it is for now.

👍 sounds good to me

Fizzadar reviewed (approved):
Awesome stuff, @vram0gh2! When properly at keyboard I’ll fix the conflicts and merge.

Fizzadar committed:
Merge branch '3.x' into filetimes_squash

Fizzadar committed:
Fix python <3.11 isoformat parsing

Fizzadar committed:
Use dates Python 3.10 can handle in tests

Fizzadar commented:
Augh damn Python 3.10 and it's useless isoformat parsing, can't believe it took until 3.11 to fix that!

Fizzadar committed:
Attempt to fix windows tests

Unfortunately I have no windows machine readily available, so this is
guesswork.
